Assets/Scripts/Controller/WaypointController.cs

```diff
@@ -1,22 +1,27 @@
-using System;
 using UnityEngine;
 
 // The waypoint controller steers the agent to the target using direct linear guidance.
-public class WaypointController : IControllerLegacy {
-  // Desired speed in m/s.
-  float _desiredSpeed;
+public class WaypointController : ControllerBase {
+  // To prevent overshooting near the waypoint, if the squared distance to the waypoint is less than
+  // the threshold, do not apply any acceleration.
+  private const float _accelerationCutoffDistanceSqr = 1.0f;
 
-  public WaypointController(Agent agent, float desiredSpeed) : base(agent) {
-    _desiredSpeed = desiredSpeed;
-  }
+  public WaypointController(IAgent agent) : base(agent) {}
+
+  // Controller-dependent implementation of the control law.
+  protected override Vector3 Plan(in Transformation relativeTransformation) {
+    Vector3 relativePosition = relativeTransformation.Position.Cartesian;
+    if (relativeTransformation.Position.Range < _accelerationCutoffDistanceSqr) {
+      return Vector3.zero;
+    }
```

Comment:
This distance check is suspicious to me - `_accelerationCutoffDistanceSqr` the name suggest a squared distance but it's being compared against Range which is a linear distance?

Reply:
Fixed, thanks!